examples/features/torch_batch_process/batch_inference/compare_with_core_api/core_api_inference.py

```diff
@@ -0,0 +1,151 @@
+
+import json
+import logging
+import math
+import os
+import pathlib
+from typing import Any, Optional
+
+import filelock
+import model
+import torch
+import torch.distributed as dist
+import torchvision as tv
+import torchvision.transforms as transforms
+from torch import nn
+
+import determined as det
+from determined import common, core, pytorch
+
+common.set_logger(False)
+
+
+def run_inference(
+    model: nn.Module,
+    data_loader: Any,
+    context: core.Context,
+    rank: int,
+    skip: int,
+    per_worker_iterate_length: int,
+) -> None:
+    model.eval()
+    records_processed = 0
+
+    inference_output_dir = "inference_out/"
+
+    # The first worker will create it, and exist_ok option makes sure subsequent workers
+    # do not run into error
+    pathlib.Path.mkdir(pathlib.Path(inference_output_dir), parents=True, exist_ok=True)
+
+    dataloader_iterator = iter(data_loader)
+
+    with torch.no_grad():
+        for batch_idx in range(skip, per_worker_iterate_length):
+            X = next(dataloader_iterator, None)
+            logging.info(f"Working on batch is {batch_idx}")
+            if X is not None:
+                data, label = X
+                output = model(data)
+                preds = output.argmax(dim=1, keepdim=True)
+
+                file_name = f"inference_out_{rank}_{batch_idx}.json"
+
+                output = []
+
+                for pred in preds:
+                    output.append(pred[0].item())
+
+                with open(
+                    os.path.join(inference_output_dir, f"{file_name}"),
+                    "w",
+                ) as f:
+                    json.dump({"predictions": output}, f)
+
+                # After each batch, synchronize and update number of catches completed
+                if context.distributed.rank == 0:
+                    work_completed_this_round = sum(context.distributed.gather(len(data)))
+                    records_processed += work_completed_this_round
+                    checkpoint_metadata = {
+                        "steps_completed": batch_idx,
+                    }
+                    with context.checkpoint.store_path(checkpoint_metadata) as (path, uuid):
+                        with open(os.path.join(path, "batch_completed.json"), "w") as file_obj:
+                            json.dump({"batch_completed": batch_idx}, file_obj)
+                else:
+                    context.distributed.gather(len(data))
+
+                if context.preempt.should_preempt():
+                    return
+
+
+def _get_data_loader(batch_size: int, total_worker: int, rank: int, skip: int) -> [Any, int]:
+    transform = transforms.Compose(
+        [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
+    )
+    with filelock.FileLock(os.path.join("/tmp", "inference.lock")):
+        inference_data = tv.datasets.CIFAR10(
+            root="/data", train=False, download=True, transform=transform
```

Comment:
Ok, I have a selfish reason for this request:

I like bind-mounting data in our examples when I have to debug them.  If I have to run a cifar example a bunch of times, literally each time I go to work on it, I write some idempotent download code that I never commit to the example (it would be a distraction in the example).

But you could definitely write an example to make it a _lot_ easier to do this hack:

- start with `os.makedir("/tmp/data", exist_ok=True)
- file lock on `os.makedir("/tmp/data/lock")`
- download to `/tmp/data/cifar`

That way, while you are developing the example, you can bind mount `/tmp/data` from the agent into the container, and you'll get the following benefits:
- the example will download data only once per agent lifetime
- the example will work even if multiple 1-slot inference jobs run simultaneously on the same agent

But at the same time, users reading the example think that the example makes perfect sense without the bind mount, which streamlines the example for readers.

Reply:
Good suggestions, thanks!